diff --git a/src/libpsl-native/CMakeLists.txt b/src/libpsl-native/CMakeLists.txt
index b6f999d..43d9aba 100644
--- a/src/libpsl-native/CMakeLists.txt
+++ b/src/libpsl-native/CMakeLists.txt
@@ -13,13 +13,4 @@ endif()
 
 set(LIBRARY_OUTPUT_PATH "${PROJECT_SOURCE_DIR}/../powershell-unix")
 
-if (CMAKE_SYSTEM_PROCESSOR MATCHES "arm*")
-    message(STATUS "Building for ARM, no tests")
-    add_subdirectory(src)
-else ()
-    # test in BUILD_DIR
-    message(STATUS "Tests enabled")
-    enable_testing()
-    add_subdirectory(src)
-    add_subdirectory(test)
-endif ()
+add_subdirectory(src)
